services/src/floodlight-modules/src/main/java/org/openkilda/floodlight/service/session/Session.java

```diff
@@ -0,0 +1,183 @@
+
+    public void resetError() {
+        error = false;
+    }
+
+    @Override
+    public void close() throws SwitchWriteException {
+        if (error) {
+            SessionRevertException e = new SessionRevertException(sw.getId());
+            incompleteRequestsStream()
+                    .forEach(entry -> entry.completeExceptionally(e));
+            return;
+        }
+
+        if (closingBarrier != null) {
+            throw new IllegalStateException("Session already closed");
+        }
+
+        OFBarrierRequest barrier = sw.getOFFactory().barrierRequest();
+        closingBarrier = prepareRequest(barrier);
+        try {
+            actualWrite(closingBarrier, barrier);
+        } catch (SwitchWriteException e) {
+            SessionCloseException closeError = new SessionCloseException(sw.getId());
+            incompleteRequestsStream()
+                    .forEach(entry -> entry.completeExceptionally(closeError));
+            throw e;
+        }
+    }
+
+    void disconnect() {
+        // must be safe to be called multiple times
+        if (completed) {
+            return;
+        }
+        completed = true;
+
+        SessionConnectionLostException e = new SessionConnectionLostException(sw.getId());
+        incompleteRequestsStream()
+                .forEach(entry -> entry.completeExceptionally(e));
+    }
+
+    boolean handleResponse(OFMessage message) {
+        CompletableFuture<Optional<OFMessage>> future;
+        future = requestsByXid.get(message.getXid());
+
+        if (future == null) {
+            throw new IllegalArgumentException(String.format(
+                    "%s must never route \"foreign\" response", group.getClass().getName()));
+        }
+        if (future.isDone()) {
+            // it can already be marked as failed by results of some session wide errors
+            return false;
+        }
+
+        if (OFType.ERROR == message.getType()) {
+            future.completeExceptionally(new SessionErrorResponseException(sw.getId(), (OFErrorMsg) message));
+        } else {
+            future.complete(Optional.of(message));
+        }
+
+        // check session completion (we have received all responses, if we got response for closing barrier request)
+        if (closingBarrier.isDone()) {
+            incompleteRequestsStream()
+                    .forEach(entry -> entry.complete(Optional.empty()));
+            return true;
+        }
+        return false;
+    }
+
+    Set<Long> getAllXids() {
+        return ImmutableSet.copyOf(requestsByXid.keySet());
+    }
+
+    private CompletableFuture<Optional<OFMessage>> prepareRequest(OFMessage message) {
+        CompletableFuture<Optional<OFMessage>> future = new CompletableFuture<>();
+
+        long xid = message.getXid();
+        requestsByXid.put(xid, future);
+        group.bindRequest(this, xid);
+
+        return future;
+    }
+
+    private void actualWrite(CompletableFuture<Optional<OFMessage>> future, OFMessage message)
```

Comment:
Actually this method doesn't need to know about the CompletableFuture instance as the logic which calls actualWrite can handle it. Moreover, the CompletableFuture doesn't require to be completed when as an exception is thrown.